File: ui/src/mixins/request.js
/*
 * Tencent is pleased to support the open source community by making 蓝鲸 available.
 * Copyright (C) 2017-2022 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mapGetters } from 'vuex'
export default {
  computed: {
    ...mapGetters('request', {
      $requestQueue: 'queue',
      $requestCache: 'cache'
    })
  },
  methods: {
    $loading(requestIds) {
      if (typeof requestIds === 'undefined') {
        return !!this.$requestQueue.length
      } if (requestIds instanceof Array) {
        return requestIds.some(requestId => this.$requestQueue.some(request => request.requestId === requestId))
      } if (typeof requestIds === 'string' && requestIds.startsWith('^=')) {
        // eslint-disable-next-line prefer-destructuring
        const requestId = requestIds.split('^=')[1]
        const matchIndex = this.$requestQueue.findIndex(request => (typeof request.requestId === 'string') && request.requestId.startsWith(requestId))
        return matchIndex !== -1
      }
      return this.$requestQueue.some(request => request.requestId === requestIds)
    }
  }
}

File: pkg/rest/request.go
/*
 * TencentBlueKing is pleased to support the open source community by making
 * 蓝鲸智云 - 配置平台 (BlueKing - CMDB) available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the MIT License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 * We undertake not to change the open source license (MIT license) applicable
 * to the current version of the project delivered to anyone in the future.
 */

package rest

import (
	"context"
	"net/http"

	"github.com/TencentBlueKing/bk-cmdb/pkg/rest/codec"
	"github.com/TencentBlueKing/bk-cmdb/pkg/validator"
)

// decodeReq ...
func decodeReq[T any](r *http.Request) (*T, error) {
	in := new(T)

	// http.Request 直接返回
	if _, ok := any(in).(*http.Request); ok {
		return any(r).(*T), nil
	}

	// 空值不需要反序列化
	if _, ok := any(in).(*EmptyReq); ok {
		return in, nil
	}

	in, err := codec.Decode[T](r)
	if err != nil {
		return nil, err
	}

	return in, nil
}

// validate 参数校验
func validateReq(ctx context.Context, req any) error {
	// http.Request 直接返回
	if _, ok := req.(*http.Request); ok {
		return nil
	}

	// 空值不需要校验
	if _, ok := req.(*EmptyReq); ok {
		return nil
	}

	return validator.Struct(ctx, req)
}
